lib/modules/datasource/maven/util.ts

```diff
@@ -91,13 +93,62 @@ export async function downloadHttpProtocol(
       // istanbul ignore next
       logger.debug({ failedUrl }, 'Unsupported host');
     } else {
-      logger.info({ failedUrl, err }, 'Unknown error');
+      logger.info({ failedUrl, err }, 'Unknown HTTP download error');
     }
     return {};
   }
 }
 
-export async function checkHttpResource(
+function isS3NotFound(err: { name: string; message: string }): boolean {
+  return err.message === 'NotFound' || err.message === 'NoSuchKey';
+}
+
+export async function downloadS3Protocol(
+  pkgUrl: URL | string
+): Promise<string> {
+  logger.trace({ url: pkgUrl.toString() }, `Attempting to load S3 dependency`);
+  let body: string;
+  try {
+    const s3Url = parseS3Url(pkgUrl.toString());
```

Comment:
The notion of `S3Url` may look confusing for the future reader. Though I can't come up with better naming: `S3Params`, `S3Coordinate`, `S3Config`? 🤷‍♂️

Reply:
I mean, while `parseS3Url` is okay as it receives URL, the result isn't URL

Reply:
So this is passed into `getObject` in the S3 client, which accepts a `GetObjectRequest`, where two of the params are `Bucket` and `Key`. We could rename `parseS3Url` to something like `getObjectRequestFromUrl` maybe?

FWIW in the v1 API of the Java AWS SDK, this class is called `AmazonS3URI`. To anyone working regularly with AWS, I don't think it's confusing, but I do get your point.

Reply:
Okay, then maybe `S3UrlParts`?